R-package/R/lgb.Dataset.R

```diff
@@ -288,6 +288,10 @@ Dataset <- R6::R6Class(
         self$set_colnames(colnames = private$colnames)
       }
 
+      # If the data didn't have feature names we take the ones defined at cpp side
+      # otherwise we just overwrite them
+      self$get_colnames()
```

Comment:
I think understand the discussion in https://github.com/microsoft/LightGBM/pull/5116/files#r843276288 that lead to this change, but I don't understand why adding this call here is necessary for the R package and I don't think it's achieving what you want (avoiding a call to the C++ side for performance reasons in `Dataset$get_feature_num_bin()`).

The use of `colnames(self)` down in this PR's proposed change to `Dataset$get_feature_num_bin()` will still make a call to the C++ side.

`colnames()` called on a `Dataset` calls `dimnames.lgb.Dataset()`, which calls `self$get_colnames()`, which gets feature names from the C++ side for a constructed Dataset.

https://github.com/microsoft/LightGBM/blob/b462d0a40f6978c0a66a47af4522c2ae12532316/R-package/R/lgb.Dataset.R#L983

If you want to avoid that call to the C++ side, I think the code in `Dataset$get_feature_num_bin()` could just reference `private$colnames` directly.

Consider the following example:

```r
library(lightgbm)
data("EuStockMarkets")

feature_names <- c("SMI", "CAC", "FTSE")
target_name <- "DAX"
stockDF <- as.data.frame(EuStockMarkets)
X_mat <- data.matrix(stockDF[, feature_names])
y <- stockDF[[target_name]]

# colnames are stored in private$colnames when initializing Dataset
dtrain <- lightgbm::lgb.Dataset(
    data = X_mat
    , label = y
    , colnames = feature_names
)
print(dtrain$.__enclos_env__$private$colnames)

# colnames are in private$colnames after construction
dtrain$construct()
print(dtrain$.__enclos_env__$private$colnames)

# set_colnames() updates the C++ side AND private$colnames
dtrain$set_colnames(c("A", "B", "C"))
print(dtrain$.__enclos_env__$private$colnames)
.Call(
    lightgbm:::LGBM_DatasetGetFeatureNames_R
    , dtrain$.__enclos_env__$private$handle
)
```

Reply:
Thank you. You're right, the call in the `Dataset$get_feature_num_bin()` should use `private$colnames`. The line here however tries to mimic the one on the python side when after constructing the dataset calls `self.feature_name = self.get_feature_name()` to make sure that the feature names on the python side are the same as the ones on the C++ side. This is for the case when no `colnames` are provided in the constructor but the default names are assigned on the C++ side and we want to retrieve the feature bins by the default names, because in that case `private$colnames` is `NULL` until we call `self$get_colnames()` once.

Reply:
> to make sure that the feature names on the python side are the same as the ones on the C++ side

Ok, I see. I still think it is not obvious, when looking at this line, that that's what is happening. It just looks like a call that got left behind, whose result isn't assigned.

Since you already know with certainty that at this point in `Dataset$construct()` that `Dataset` has been constructed, I think it would be clearer to directly set `private$colnames` like this:

```r
private$colnames <- .Call(
    LGBM_DatasetGetFeatureNames_R
    , private$handle
)
```

Instead of relying on the fact that `self$get_colnames()` happens to do that while it's also getting the features from the C++ side.

Looking at this again, I also recommend reformatting this comment to the following:

```text
# Ensure that private$colnames matches the feature names on the C++ side. This line is necessary
# in cases like constructing from a file or from a matrix with no column names.
```

So that it's a bit clearer when this can happen.

Reply:
Done in 302a6a1